src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -0,0 +1,327 @@
+        internal JsonTypeInfo GetOrAddJsonTypeInfoForRootType(Type type)
+        {
+            JsonTypeInfo? jsonTypeInfo = _lastTypeInfo;
+
+            if (jsonTypeInfo?.Type != type)
+            {
+                jsonTypeInfo = GetOrAddJsonTypeInfo(type);
+                _lastTypeInfo = jsonTypeInfo;
+            }
+
+            return jsonTypeInfo;
+        }
+
+        internal void ClearCaches()
+        {
+            _cachingContext?.Clear();
+            _lastTypeInfo = null;
+        }
+
+        private void InitializeCachingContext()
+        {
+            _cachingContext = TrackedCachingContexts.GetOrCreate(this);
+        }
+
+        /// <summary>
+        /// Stores and manages all reflection caches for one or more <see cref="JsonSerializerOptions"/> instances.
+        /// NB the type encapsulates the original options instance and only consults that one when building new types;
+        /// this is to prevent multiple options instances from leaking into the object graphs of converters which
+        /// could break user invariants.
+        /// </summary>
+        private sealed class CachingContext
+        {
+            private readonly ConcurrentDictionary<Type, JsonConverter> _converterCache = new();
+            private readonly ConcurrentDictionary<Type, JsonTypeInfo> _jsonTypeInfoCache = new();
+
+            public CachingContext(JsonSerializerOptions options)
+            {
+                Options = options;
+                _ = Count;
+            }
+
+            public JsonSerializerOptions Options { get; }
+            public int Count => _converterCache.Count + _jsonTypeInfoCache.Count;
+            public JsonConverter GetOrAddConverter(Type type) => _converterCache.GetOrAdd(type, Options.GetConverterFromType);
+            public JsonTypeInfo GetOrAddJsonTypeInfo(Type type) => _jsonTypeInfoCache.GetOrAdd(type, Options.GetJsonTypeInfoFromContextOrCreate);
+            public bool TryGetJsonTypeInfo(Type type, [NotNullWhen(true)] out JsonTypeInfo? typeInfo) => _jsonTypeInfoCache.TryGetValue(type, out typeInfo);
+            public bool IsJsonTypeInfoCached(Type type) => _jsonTypeInfoCache.ContainsKey(type);
+
+            public void Clear()
+            {
+                _converterCache.Clear();
+                _jsonTypeInfoCache.Clear();
+            }
+        }
+
+        /// <summary>
+        /// Defines a cache of CachingContexts; instead of using a ConditionalWeakTable which can be slow to traverse
+        /// this approach uses a regular dictionary pointing to weak references of <see cref="CachingContext"/>.
+        /// Relevant caching contexts are looked up using the equality comparison defined by <see cref="EqualityComparer"/>.
+        /// </summary>
+        private static class TrackedCachingContexts
+        {
+            private const int MaxTrackedContexts = 64;
+            private static readonly ConcurrentDictionary<JsonSerializerOptions, WeakReference<CachingContext>> s_cache =
```

Comment:
I think this should make the cache work reasonably well with collectible assemblies, but it's hard to tell for sure. Could you please add a test which:
* Creates a new unloadable ALC and loads some assembly with test types into it
* Serializes/deserializes objects of the types from the assembly above
* Unloads the ALC
* Validates that the ALC was actually unloaded

Good guide what to do is here: https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability

In short what we want to avoid:
Global GC root which holds onto anything from unloadable ALCs as that will prevent the ALC from unloading. We already have quite a few caches in the FX which do this, so let's not add another one.

Reply:
> In short what we want to avoid:
Global GC root which holds onto anything from unloadable ALCs as that will prevent the ALC from unloading. We already have quite a few caches in the FX which do this, so let's not add another one.

System.Text.Json already suffers from that issue since we're rooting the default JsonSerializerOptions instance (and any caches it may create):

https://github.com/dotnet/runtime/blob/cff924f9c823208b21d85f8fd5eed7eee8d4017a/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs#L35

We need to work out an eviction policy that takes into account the rooted instance as well, however that's not in scope for the current PR, which only addresses the performance issue.

Reply:
Do we have a bug tracking this somewhere? (what I meant is that we should have a bug :wink:)